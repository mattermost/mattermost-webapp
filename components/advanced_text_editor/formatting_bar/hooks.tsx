// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useEffect, useLayoutEffect, useState} from 'react';
import {Instance} from '@popperjs/core';

import {debounce} from 'lodash';

import {MarkdownMode} from 'utils/markdown/apply_markdown';

type WideMode = 'wide' | 'normal' | 'narrow';

export function useGetLatest<T>(val: T) {
    const ref = React.useRef<T>(val);
    ref.current = val;
    return React.useCallback(() => ref.current, []);
}

const useResponsiveFormattingBar = (ref: React.RefObject<HTMLDivElement>): WideMode => {
    const [wideMode, setWideMode] = useState<WideMode>('wide');
    const handleResize = debounce(() => {
        if (ref.current?.clientWidth === undefined) {
            return;
        }

        if (ref.current.clientWidth > 640) {
            setWideMode('wide');
        }

        if (ref.current.clientWidth >= 424 && ref.current.clientWidth <= 640) {
            setWideMode('normal');
        }

        if (ref.current.clientWidth < 424) {
            setWideMode('narrow');
        }
    }, 10);

    useLayoutEffect(() => {
        if (!ref.current) {
            return () => {};
        }

        let sizeObserver: ResizeObserver | null = new ResizeObserver(handleResize);

        sizeObserver.observe(ref.current);

        return () => {
            sizeObserver!.disconnect();
            sizeObserver = null;
        };
    }, [handleResize, ref]);

    return wideMode;
};

const MAP_WIDE_MODE_TO_CONTROLS_QUANTITY: {[key in WideMode]: number} = {
    wide: 9,
    normal: 5,
    narrow: 3,
};

export const useFormattingBarControls = (
    formattingBarRef: React.RefObject<HTMLDivElement>,
): {
    controls: MarkdownMode[];
    hiddenControls: MarkdownMode[];
    wideMode: WideMode;
} => {
    const wideMode = useResponsiveFormattingBar(formattingBarRef);

    const allControls: MarkdownMode[] = ['bold', 'italic', 'strike', 'heading', 'link', 'code', 'quote', 'ul', 'ol'];

    const controlsLength = MAP_WIDE_MODE_TO_CONTROLS_QUANTITY[wideMode];

    const controls = allControls.slice(0, controlsLength);
    const hiddenControls = allControls.slice(controlsLength);

    return {
        controls,
        hiddenControls,
        wideMode,
    };
};

export const useUpdateOnVisibilityChange = (update: Instance['update'] | null, isVisible: boolean) => {
    const updateComponent = async () => {
        if (!update) {
            return;
        }
        await update();
    };

    useEffect(() => {
        if (!isVisible) {
            return;
        }
        updateComponent();
    }, [isVisible]);
};
