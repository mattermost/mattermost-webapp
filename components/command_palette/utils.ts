// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Channel} from '@mattermost/types/channels';
import {Team} from '@mattermost/types/teams';

// See LICENSE.txt for license information.
import {CommandPaletteItem} from './command_palette_list_item/command_palette_list_item';
import {
    CmdPalettePictographIcon,
    CmdPalettePictographType,
} from './command_palette_list_item/command_palette_list_item_pictograph';
import {ChannelType, CommandPaletteEntities} from './types';

const channelTypeOrder = {
    [ChannelType.OPEN_CHANNEL]: 0,
    [ChannelType.PRIVATE_CHANNEL]: 1,
    [ChannelType.DM_CHANNEL]: 2,
    [ChannelType.GM_CHANNEL]: 3,
};

function filterName(name: string): string {
    return name.replace(/[.,'"\/#!$%\^&\*;:{}=\-_`~()]/g, ''); // eslint-disable-line no-useless-escape
}

export function channelToCommandPaletteItemTransformer(channels: Channel[], teams: Record<string, Team>): CommandPaletteItem[] {
    return channels.map((channel) => {
        const isArchived = channel.delete_at ? channel.delete_at !== 0 : false;

        let channelIcon = CmdPalettePictographIcon.GLOBE;
        if (isArchived) {
            channelIcon = CmdPalettePictographIcon.ARCHIVE;
        }
        return {
            description: channel.type === ChannelType.DM_CHANNEL ? channel.name : '',
            id: channel.id,
            isDeactivated: isArchived,
            teamName: teams[channel.team_id]?.name,
            pictograph: {
                type: CmdPalettePictographType.ICON,
                pictographItem: channelIcon,
            },
            subType: channel.type as ChannelType,
            title: channel.display_name,
            type: CommandPaletteEntities.Channel,
            isArchived,
        };
    });
}

function sortChannelsByTypeAndDisplayName(locale: string, cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem): number {
    if (channelTypeOrder[cmdPaletteItemA.subType] !== channelTypeOrder[cmdPaletteItemB.subType]) {
        if (channelTypeOrder[cmdPaletteItemA.subType] < channelTypeOrder[cmdPaletteItemB.subType]) {
            return -1;
        }

        return 1;
    }

    const aDisplayName = filterName(cmdPaletteItemA.title);
    const bDisplayName = filterName(cmdPaletteItemB.title);

    if (aDisplayName !== bDisplayName) {
        return aDisplayName.toLowerCase().localeCompare(bDisplayName.toLowerCase(), locale, {numeric: true});
    }

    return cmdPaletteItemA.title.toLowerCase().localeCompare(cmdPaletteItemA.title.toLowerCase(), locale, {numeric: true});
}

function sortChannelsByRecencyAndTypeAndDisplayName(cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem) {
    if (cmdPaletteItemA.lastViewedAt && cmdPaletteItemB.lastViewedAt) {
        return cmdPaletteItemB.lastViewedAt - cmdPaletteItemA.lastViewedAt;
    } else if (cmdPaletteItemA.lastViewedAt) {
        return -1;
    } else if (cmdPaletteItemB.lastViewedAt) {
        return 1;
    }

    // MM-12677 When this is migrated this needs to be fixed to pull the user's locale
    return sortChannelsByTypeAndDisplayName('en', cmdPaletteItemA, cmdPaletteItemB);
}

export function commandPaletteSorter(prefix: string, cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem) {
    if (cmdPaletteItemA.isArchived && !cmdPaletteItemB.isArchived) {
        return 1;
    } else if (!cmdPaletteItemA.isArchived && cmdPaletteItemB.isArchived) {
        return -1;
    }

    if (cmdPaletteItemA.isDeactivated && !cmdPaletteItemB.isDeactivated) {
        return 1;
    } else if (cmdPaletteItemB.isDeactivated && !cmdPaletteItemA.isDeactivated) {
        return -1;
    }

    let aDisplayName = cmdPaletteItemA.title.toLowerCase();
    let bDisplayName = cmdPaletteItemA.title.toLowerCase();

    if (cmdPaletteItemA.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemA.subType === ChannelType.DM_CHANNEL &&
        aDisplayName.startsWith('@')) {
        aDisplayName = aDisplayName.substring(1);
    }

    if (cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemB.subType === ChannelType.DM_CHANNEL && bDisplayName.startsWith('@')) {
        bDisplayName = bDisplayName.substring(1);
    }

    const aStartsWith = aDisplayName.startsWith(prefix) || cmdPaletteItemA.description.toLowerCase().startsWith(prefix);
    const bStartsWith = bDisplayName.startsWith(prefix) || cmdPaletteItemB.description.toLowerCase().startsWith(prefix);

    // Open channels user haven't interacted should be at the  bottom of the list
    if (cmdPaletteItemA.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemA.subType === ChannelType.OPEN_CHANNEL &&
        !cmdPaletteItemA.lastViewedAt &&
        ((cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
            cmdPaletteItemB.subType !== ChannelType.OPEN_CHANNEL) || cmdPaletteItemB.lastViewedAt)) {
        return 1;
    } else if ((cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemB.subType === ChannelType.OPEN_CHANNEL) || !cmdPaletteItemB.lastViewedAt) {
        return -1;
    }

    // Sort channels starting with the search term first
    if (aStartsWith && !bStartsWith) {
        return -1;
    } else if (!aStartsWith && bStartsWith) {
        return 1;
    }
    return sortChannelsByRecencyAndTypeAndDisplayName(cmdPaletteItemA, cmdPaletteItemB);
}
