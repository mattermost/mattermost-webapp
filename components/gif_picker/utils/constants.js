// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export default {
    ItemTapAction: {
        OPEN_EMBED_PAGE: 1,
        SHARE: 2,
    },

    Tab: {
        TRENDING: 0,
        REACTIONS: 1,
    },

    appName: {
        mattermost: 'mattermost',
    },
};
