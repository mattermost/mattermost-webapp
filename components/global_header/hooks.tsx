// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {MutableRefObject, useEffect, useRef} from 'react';
import {useSelector} from 'react-redux';

import {getCurrentUser, isFirstAdmin, isCurrentUserSystemAdmin} from 'mattermost-redux/selectors/entities/users';
import {UserProfile} from '@mattermost/types/users';
import {isModalOpen} from 'selectors/views/modals';
import {GlobalState} from 'types/store';

/**
 * Hook that alerts clicks outside of the passed ref.
 */
export function useClickOutsideRef(ref: MutableRefObject<HTMLElement | null>, handler: () => void): void {
    useEffect(() => {
        function onMouseDown(event: MouseEvent) {
            const target = event.target as any;
            if (ref.current && target instanceof Node && !ref.current.contains(target)) {
                handler();
            }
        }

        // Bind the event listener
        document.addEventListener('mousedown', onMouseDown);
        return () => {
            // Unbind the event listener on clean up
            document.removeEventListener('mousedown', onMouseDown);
        };
    }, [ref, handler]);
}

export const useFirstAdminUser = (): boolean => {
    return useSelector(isFirstAdmin);
};

export const useIsCurrentUserSystemAdmin = (): boolean => {
    return useSelector(isCurrentUserSystemAdmin);
};

export const useIsLoggedIn = (): boolean => {
    return Boolean(useSelector<GlobalState, UserProfile>(getCurrentUser));
};

/**
 * Hook that returns the current open state of the specified modal
 * - returns both the direct boolean for regular use and a ref that contains the boolean for usage in a callback
 */
export const useIsModalOpen = (modalIdentifier: string): [boolean, React.RefObject<boolean>] => {
    const modalOpenState = useSelector((state: GlobalState) => isModalOpen(state, modalIdentifier));
    const modalOpenStateRef = useRef(modalOpenState);

    useEffect(() => {
        modalOpenStateRef.current = modalOpenState;
    }, [modalOpenState]);

    return [modalOpenState, modalOpenStateRef];
};
