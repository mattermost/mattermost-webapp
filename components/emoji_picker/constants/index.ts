// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {EmojiCategory} from '@mattermost/types/emojis';

import * as Emoji from 'utils/emoji';

import {Category, Categories} from '../types';

export const RECENT = 'recent';
export const SEARCH_RESULTS = 'searchResults';
export const SMILEY_EMOTION = 'smileys-emotion';
export const CUSTOM = 'custom';

const categoryClass: Map<EmojiCategory, string> = new Map([
    [RECENT, 'icon-clock-outline'],
    [SMILEY_EMOTION, 'icon-emoticon-happy-outline'],
    ['people-body', 'icon-account-outline'],
    ['animals-nature', 'icon-leaf-outline'],
    ['food-drink', 'icon-food-apple'],
    ['activities', 'icon-basketball'],
    ['travel-places', 'icon-airplane-variant'],
    ['objects', 'icon-lightbulb-outline'],
    ['symbols', 'icon-heart-outline'],
    ['flags', 'icon-flag-outline'],
    [CUSTOM, 'icon-emoticon-custom-outline'],
    [SEARCH_RESULTS, ''],
]);

function createCategory(name: EmojiCategory): Category {
    return {
        name,
        id: Emoji.CategoryTranslations.get(name) || '',
        className: categoryClass.get(name) || '',
        message: Emoji.CategoryMessage.get(name)!,
    };
}

export const RECENT_EMOJI_CATEGORY: Pick<Categories, 'recent'> = {recent: createCategory(RECENT)};
export const SEARCH_EMOJI_CATEGORY: Pick<Categories, typeof SEARCH_RESULTS> = {searchResults: createCategory(SEARCH_RESULTS)};

export const CATEGORIES: Categories = Emoji.CategoryNames.
    filter((category) => !(category === 'recent' || category === 'searchResults')).
    reduce((previousCategory, currentCategory) => {
        return {
            ...previousCategory,
            [currentCategory]: createCategory(currentCategory as EmojiCategory),
        };
    }, {} as Categories);

export const EMOJI_PER_ROW = 9; // needs to match variable `$emoji-per-row` in _variables.scss
export const ITEM_HEIGHT = 36; //as per .emoji-picker__item height in _emoticons.scss
export const EMOJI_CONTAINER_HEIGHT = 290; // If this changes, the spaceRequiredAbove and spaceRequiredBelow props passed to the EmojiPickerOverlay must be updated

export const CATEGORY_HEADER_ROW = 'categoryHeaderRow';
export const EMOJIS_ROW = 'emojisRow';

export const EMOJI_SCROLL_THROTTLE_DELAY = 150;
export const EMOJI_ROWS_OVERSCAN_COUNT = 1;

export const CUSTOM_EMOJIS_PER_PAGE = 200;
export const CUSTOM_EMOJI_SEARCH_THROTTLE_TIME_MS = 1000;
