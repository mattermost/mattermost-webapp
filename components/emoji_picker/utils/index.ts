// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Emoji, EmojiCategory} from 'mattermost-redux/types/emojis';

import {compareEmojis} from 'utils/emoji_utils';

import {Categories} from '../types';

function sortEmojis(emojis: Emoji[], recentEmojiStrings: string[], filter: string) {
    const recentEmojis: Emoji[] = [];
    const emojisMinusRecent: Emoji[] = [];

    Object.values(emojis).forEach((emoji) => {
        let emojiArray = emojisMinusRecent;
        const alias = 'short_names' in emoji ? emoji.short_names : [emoji.name];
        for (let i = 0; i < alias.length; i++) {
            if (recentEmojiStrings.includes(alias[i].toLowerCase())) {
                emojiArray = recentEmojis;
            }
        }

        emojiArray.push(emoji);
    });

    return [
        ...recentEmojis.sort((firstEmoji, secondEmoji) => compareEmojis(firstEmoji, secondEmoji, filter)),
        ...emojisMinusRecent.sort((firstEmoji, secondEmoji) => compareEmojis(firstEmoji, secondEmoji, filter)),
    ];
}

function getFilteredEmojis(allEmojis: Record<string, Emoji>, filter: string, recentEmojisString: string[]): Emoji[] {
    const emojis = Object.values(allEmojis).filter((emoji) => {
        const alias = 'short_names' in emoji ? emoji.short_names : [emoji.name];
        for (let i = 0; i < alias.length; i++) {
            if (alias[i].toLowerCase().includes(filter)) {
                return true;
            }
        }

        return false;
    });

    return sortEmojis(emojis, recentEmojisString, filter);
}

export function getEmojisByCategory(allEmojis: Record<string, Emoji>, categories: Categories, categoryName: EmojiCategory, filter: string, recentEmojisString: string[]): Emoji[] {
    if (filter.length) {
        return getFilteredEmojis(allEmojis, filter, recentEmojisString);
    }

    // if (categoryName === 'recent') {
    //     const recentEmojiIds = this.state.categories?.recent?.emojiIds ?? [];
    //     if (recentEmojiIds.length === 0) {
    //         return [];
    //     }

    //     const recentEmojis = new Map();

    //     recentEmojiIds.forEach((emojiId) => {
    //         const emoji = this.state.allEmojis[emojiId];
    //         const emojiSkin = getSkin(emoji);

    //         if (emojiSkin) {
    //             const emojiWithUserSkin = this.convertEmojiToUserSkinTone(emoji, emojiSkin, this.props.userSkinTone);
    //             recentEmojis.set(emojiWithUserSkin.unified, emojiWithUserSkin);
    //         } else {
    //             recentEmojis.set(emojiId, emoji);
    //         }
    //     });

    //     return Array.from(recentEmojis.values());
    // }

    return categories[categoryName].emojiIds.map((emojiId) => allEmojis[emojiId]);
}
