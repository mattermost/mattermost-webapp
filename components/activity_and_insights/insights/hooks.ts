// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {useSelector} from 'react-redux';

import {getCloudSubscription, getSubscriptionProduct} from 'mattermost-redux/selectors/entities/cloud';
import {getConfig, getLicense} from 'mattermost-redux/selectors/entities/general';

import {isCloudLicense} from 'utils/license_utils';
import {CloudProducts, InsightsScopes} from 'utils/constants';

import {setGlobalItem} from 'actions/storage';

import {useGlobalState} from 'stores/hooks';

/**
 * Returns some checks for free trial users or starter licenses
 */
export function useLicenseChecks(): {isStarterFree: boolean; isFreeTrial: boolean; isEnterpriseReady: boolean} {
    const subscription = useSelector(getCloudSubscription);
    const license = useSelector(getLicense);
    const subscriptionProduct = useSelector(getSubscriptionProduct);
    const config = useSelector(getConfig);

    const isCloud = isCloudLicense(license);
    const isCloudStarterFree = isCloud && subscriptionProduct?.sku === CloudProducts.STARTER;
    const isCloudFreeTrial = isCloud && subscription?.is_free_trial === 'true';

    const isEnterpriseReady = config.BuildEnterpriseReady === 'true';
    const isSelfHostedStarter = isEnterpriseReady && (license.IsLicensed === 'false');
    const isSelfHostedFreeTrial = license.IsTrial === 'true';

    const isStarterFree = isCloudStarterFree || isSelfHostedStarter;
    const isFreeTrial = isCloudFreeTrial || isSelfHostedFreeTrial;

    return {
        isStarterFree,
        isFreeTrial,
        isEnterpriseReady,
    };
}

export function useGetFilterType(): [string, (value: string) => ReturnType<typeof setGlobalItem>] {
    const {isStarterFree, isEnterpriseReady} = useLicenseChecks();
    const [filterType, setFilterType] = useGlobalState(InsightsScopes.TEAM, 'insightsScope');

    let returnedFilterType = filterType;

    // Force MY Insights for free trial users
    if (isStarterFree || !isEnterpriseReady) {
        returnedFilterType = InsightsScopes.MY;
    }

    return [
        returnedFilterType,
        setFilterType,
    ];
}
