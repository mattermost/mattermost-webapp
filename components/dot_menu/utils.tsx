// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {trackEvent} from 'actions/telemetry_actions';

import {EventTypes, TELEMETRY_CATEGORIES} from 'utils/constants';

export type ChangeEvent = React.KeyboardEvent | React.MouseEvent;

export const trackDotMenuEvent = (e: ChangeEvent, suffix: string): void => {
    if (e.type === EventTypes.CLICK) {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.CLICK + '_' + suffix);
    } else {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.SHORTCUT + '_ ' + suffix);
    }
};
