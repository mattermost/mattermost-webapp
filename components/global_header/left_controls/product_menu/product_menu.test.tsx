// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';
import {shallow} from 'enzyme';

import {TopLevelProducts} from 'utils/constants';
import {TestHelper} from 'utils/test_helper';

import * as hooks from '../../hooks';

import ProductMenu, {ProductMenuButton, ProductMenuContainer} from './product_menu';
import ProductMenuItem from './product_menu_item';
import ProductMenuList from './product_menu_list';

const spyProduct = jest.spyOn(hooks, 'useCurrentProductId');
spyProduct.mockReturnValue(null);

describe('components/global/product_switcher', () => {
    beforeEach(() => {
        const products = [
            TestHelper.makeProduct(TopLevelProducts.BOARDS),
            TestHelper.makeProduct(TopLevelProducts.PLAYBOOKS),
        ];
        const spyProducts = jest.spyOn(hooks, 'useProducts');
        spyProducts.mockReturnValue(products);
    });

    it('should match snapshot', () => {
        const wrapper = shallow(<ProductMenu/>);

        expect(wrapper).toMatchSnapshot();
    });

    it('should render once when there are no top level products available', () => {
        const wrapper = shallow(<ProductMenu/>);

        const spyProducts = jest.spyOn(hooks, 'useProducts');

        spyProducts.mockReturnValue([]);
        expect(wrapper.find(ProductMenuItem).at(0)).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render the correct amount of times when there are products available', () => {
        const wrapper = shallow(<ProductMenu/>);
        const products = [
            TestHelper.makeProduct(TopLevelProducts.BOARDS),
            TestHelper.makeProduct(TopLevelProducts.PLAYBOOKS),
        ];

        const spyProducts = jest.spyOn(hooks, 'useProducts');
        spyProducts.mockReturnValue(products);

        expect(wrapper.find(ProductMenuItem)).toHaveLength(3);
        expect(wrapper).toMatchSnapshot();
    });

    it('should have an active button state when the switcher menu is open', () => {
        const wrapper = shallow(<ProductMenu/>);
        const setState = jest.fn();

        const useStateSpy = jest.spyOn(React, 'useState');
        useStateSpy.mockImplementation(() => [false, setState]);

        wrapper.find(ProductMenuContainer).simulate('click');
        expect(wrapper.find(ProductMenuButton).props().active).toEqual(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot with product switcher menu', () => {
        const wrapper = shallow(<ProductMenu/>);

        expect(wrapper.find(ProductMenuList)).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });
});
