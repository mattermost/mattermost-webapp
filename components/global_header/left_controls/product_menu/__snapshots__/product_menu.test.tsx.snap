// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/global/product_switcher should have an active button state when the switcher menu is open 1`] = `
<div>
  <MenuWrapper
    animationComponent={[Function]}
    className=""
    open={true}
  >
    <ProductMenuContainer
      onClick={[Function]}
    >
      <ProductMenuButton
        active={true}
        aria-label="Select to open product switch menu."
      />
      <Connect(ProductMenuTip) />
      <ProductBranding />
    </ProductMenuContainer>
    <Menu
      ariaLabel="switcherOpen"
      className="product-switcher-menu"
    >
      <ProductMenuItem
        active={true}
        destination="/"
        icon="product-channels"
        onClick={[Function]}
        text="Channels"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-boards"
        key="Boards"
        onClick={[Function]}
        text="Boards"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-playbooks"
        key="Playbooks"
        onClick={[Function]}
        text="Playbooks"
      />
      <Connect(ProductMenuList)
        isMessaging={true}
        onClick={[Function]}
      />
    </Menu>
  </MenuWrapper>
</div>
`;

exports[`components/global/product_switcher should match snapshot 1`] = `
<div>
  <MenuWrapper
    animationComponent={[Function]}
    className=""
    open={false}
  >
    <ProductMenuContainer
      onClick={[Function]}
    >
      <ProductMenuButton
        active={false}
        aria-label="Select to open product switch menu."
      />
      <Connect(ProductMenuTip) />
      <ProductBranding />
    </ProductMenuContainer>
    <Menu
      ariaLabel="switcherOpen"
      className="product-switcher-menu"
    >
      <ProductMenuItem
        active={true}
        destination="/"
        icon="product-channels"
        onClick={[Function]}
        text="Channels"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-boards"
        key="Boards"
        onClick={[Function]}
        text="Boards"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-playbooks"
        key="Playbooks"
        onClick={[Function]}
        text="Playbooks"
      />
      <Connect(ProductMenuList)
        isMessaging={true}
        onClick={[Function]}
      />
    </Menu>
  </MenuWrapper>
</div>
`;

exports[`components/global/product_switcher should match snapshot with product switcher menu 1`] = `
<div>
  <MenuWrapper
    animationComponent={[Function]}
    className=""
    open={false}
  >
    <ProductMenuContainer
      onClick={[Function]}
    >
      <ProductMenuButton
        active={false}
        aria-label="Select to open product switch menu."
      />
      <Connect(ProductMenuTip) />
      <ProductBranding />
    </ProductMenuContainer>
    <Menu
      ariaLabel="switcherOpen"
      className="product-switcher-menu"
    >
      <ProductMenuItem
        active={true}
        destination="/"
        icon="product-channels"
        onClick={[Function]}
        text="Channels"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-boards"
        key="Boards"
        onClick={[Function]}
        text="Boards"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-playbooks"
        key="Playbooks"
        onClick={[Function]}
        text="Playbooks"
      />
      <Connect(ProductMenuList)
        isMessaging={true}
        onClick={[Function]}
      />
    </Menu>
  </MenuWrapper>
</div>
`;

exports[`components/global/product_switcher should render once when there are no top level products available 1`] = `
<div>
  <MenuWrapper
    animationComponent={[Function]}
    className=""
    open={false}
  >
    <ProductMenuContainer
      onClick={[Function]}
    >
      <ProductMenuButton
        active={false}
        aria-label="Select to open product switch menu."
      />
      <Connect(ProductMenuTip) />
      <ProductBranding />
    </ProductMenuContainer>
    <Menu
      ariaLabel="switcherOpen"
      className="product-switcher-menu"
    >
      <ProductMenuItem
        active={true}
        destination="/"
        icon="product-channels"
        onClick={[Function]}
        text="Channels"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-boards"
        key="Boards"
        onClick={[Function]}
        text="Boards"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-playbooks"
        key="Playbooks"
        onClick={[Function]}
        text="Playbooks"
      />
      <Connect(ProductMenuList)
        isMessaging={true}
        onClick={[Function]}
      />
    </Menu>
  </MenuWrapper>
</div>
`;

exports[`components/global/product_switcher should render the correct amount of times when there are products available 1`] = `
<div>
  <MenuWrapper
    animationComponent={[Function]}
    className=""
    open={false}
  >
    <ProductMenuContainer
      onClick={[Function]}
    >
      <ProductMenuButton
        active={false}
        aria-label="Select to open product switch menu."
      />
      <Connect(ProductMenuTip) />
      <ProductBranding />
    </ProductMenuContainer>
    <Menu
      ariaLabel="switcherOpen"
      className="product-switcher-menu"
    >
      <ProductMenuItem
        active={true}
        destination="/"
        icon="product-channels"
        onClick={[Function]}
        text="Channels"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-boards"
        key="Boards"
        onClick={[Function]}
        text="Boards"
      />
      <ProductMenuItem
        active={false}
        destination=""
        icon="product-playbooks"
        key="Playbooks"
        onClick={[Function]}
        text="Playbooks"
      />
      <Connect(ProductMenuList)
        isMessaging={true}
        onClick={[Function]}
      />
    </Menu>
  </MenuWrapper>
</div>
`;
